lighthouse-core/config/index.js

```diff
@@ -0,0 +1,161 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const defaultConfig = require('./default.json');
+const recordsFromLogs = require('../lib/network-recorder').recordsFromLogs;
+const CriticalRequestChainsGatherer = require('../driver/gatherers/critical-request-chains');
+const Driver = require('../driver');
+const log = require('../lib/log');
+
+class ConfigParser {
+  static filterAudits(audits, whitelist) {
+    // It's possible we didn't get given any audits (but existing audit results), in which case
+    // there is no need to do any filter work.
+    if (!audits) {
+      return;
+    }
+
+    const rejected = [];
+    const filteredAudits = audits.filter(a => {
+      // If there is no whitelist, assume all.
+      if (!whitelist) {
+        return true;
+      }
+
+      const auditName = a.toLowerCase();
+      const inWhitelist = whitelist.has(auditName);
+
+      if (!inWhitelist) {
+        rejected.push(auditName);
+      }
+
+      return inWhitelist;
+    });
+
+    if (rejected.length) {
+      log.log('info', 'Running these audits:', `${filteredAudits.join(', ')}`);
+      log.log('info', 'Skipping these audits:', `${rejected.join(', ')}`);
+    }
+
+    return filteredAudits;
+  }
+
+  static expandAudits(audits) {
+    // It's possible we didn't get given any audits (but existing audit results), in which case
+    // there is no need to do any expansion work.
+    if (!audits) {
+      return;
+    }
+
+    return audits.map(audit => {
+      // If this is already instantiated, don't do anything else.
+      if (typeof audit !== 'string') {
+        return audit;
+      }
+
+      try {
+        return require(`../audits/${audit}`);
+      } catch (requireError) {
+        throw new Error(`Unable to locate audit: ${audit}`);
+      }
+    });
+  }
+
+  static getGatherersNeededByAudits(config) {
+    // It's possible we didn't get given any audits (but existing audit results), in which case
+    // there is no need to do any work here.
+    if (!config.audits) {
+      return new Set();
+    }
+
+    return config.audits.reduce((list, audit) => {
+      audit.meta.requiredArtifacts.forEach(artifact => list.add(artifact));
+      return list;
+    }, new Set());
+  }
+
+  static filterPasses(config) {
+    const requiredGatherers = ConfigParser.getGatherersNeededByAudits(config);
+
+    // Make sure we only have the gatherers that are needed by the audits
+    // that have been listed in the config.
+    const filteredPasses = config.passes.map(pass => {
+      pass.gatherers = pass.gatherers.filter(gatherer => {
+        if (typeof gatherer !== 'string') {
+          return requiredGatherers.has(gatherer.name);
+        }
+
+        try {
+          const GathererClass = Driver.getGathererClass(gatherer);
```

Comment:
I wonder if `getGathererClass()` should live here in `ConfigParser` and not in `Driver`. To me it makes sense for it to know how to resolve gatherers since it knows how to resolve audits. OTOH, there is an encapsulation argument: driver/index.js handles how to find gatherers, but in that case it feels like gather filtering should be happening in `Driver.run`